financial-templates-lib/clients/ExpiringMultiPartyClient.js

```diff
@@ -83,37 +83,59 @@ class ExpiringMultiPartyClient {
 
   // Returns an array of sponsor addresses.
   getAllSponsors() {
-    return this.sponsorAddresses;
+    return this.activeSponsors;
   }
 
   // Returns the last update timestamp.
   getLastUpdateTime() {
     return this.lastUpdateTimestamp;
   }
 
-  async update() {
-    // Fetch contract state variables in parallel.
-    const [collateralRequirement, liquidationLiveness, events, cumulativeFeeMultiplier] = await Promise.all([
+  async initialSetup() {
+    const [collateralRequirement, liquidationLiveness, cumulativeFeeMultiplier] = await Promise.all([
       this.emp.methods.collateralRequirement().call(),
       this.emp.methods.liquidationLiveness().call(),
-      this.emp.getPastEvents("NewSponsor", { fromBlock: 0 }),
       this.emp.methods.cumulativeFeeMultiplier().call()
     ]);
     this.collateralRequirement = this.toBN(collateralRequirement.toString());
     this.liquidationLiveness = Number(liquidationLiveness);
-    this.sponsorAddresses = [...new Set(events.map(e => e.returnValues.sponsor))];
     this.cumulativeFeeMultiplier = this.toBN(cumulativeFeeMultiplier.toString());
+  }
+
+  async update() {
+    // If it is the first run then get contract contestants. This only needs to be called once.
+    if (!this.collateralRequirement) {
+      await this.initialSetup();
+    }
+    // Fetch contract state variables in parallel.
+    const [newSponsorEvents, endedSponsorEvents, liquidationCreatedEvents, currentTime] = await Promise.all([
```

Comment:
Drive-by, why are you querying these events now? Shouldn't these be handled by the EMPEventClient (which also needs refactoring)?

Reply:
take a look at the implementation below. I use the two sets of events to build an array of `ActiveSponsors`. I use the `EndedSponsorPosition` to filter out sponsors that have ended their position from the `NewSponsor` events. This means we don't have to query `this.emp.methods.positions` for sponsors that have ended their positions.